File: templates/project/lib/assets.ts
/**
 * Environment-Aware Asset Helper for TypeScript/JavaScript Projects
 *
 * This utility provides automatic asset URL resolution based on environment:
 * - Development: Uses local file paths
 * - Production: Uses CDN URLs
 * - Test: Configurable behavior
 *
 * @module assets
 * @version 1.0.0
 * @license MIT
 */

/**
 * Asset resolution mode
 * - 'local': Always use local paths
 * - 'cdn': Always use CDN URLs
 * - 'auto': Automatic based on environment (default)
 */
export type AssetMode = 'local' | 'cdn' | 'auto';

/**
 * Environment-specific asset resolution strategy
 * - 'cdn-production-local-dev': CDN in production, local in development (default)
 * - 'cdn-always': Always use CDN URL regardless of environment
 * - 'local-always': Always use local path regardless of environment
 */
export type EnvMode = 'cdn-production-local-dev' | 'cdn-always' | 'local-always';

/**
 * Detected environment type
 */
type Environment = 'development' | 'production' | 'test';

/**
 * Asset URL resolution configuration
 */
interface AssetConfig {
  localPath: string;
  cdnUrl: string;
  envMode?: EnvMode;
}

/**
 * AssetResolver - Singleton class for environment-aware asset URL resolution
 *
 * Provides cached environment detection and URL resolution logic.
 *
 * @example
 * ```typescript
 * const resolver = AssetResolver.getInstance();
 * const url = resolver.getAssetUrl('/media/logo.png', 'https://cdn.example.com/logos/logo.png');
 * ```
 */
export class AssetResolver {
  private static instance: AssetResolver | null = null;
  private environment: Environment;
  private assetMode: AssetMode;

  /**
   * Private constructor - use getInstance() instead
   */
  private constructor() {
    this.environment = this.detectEnvironment();
    this.assetMode = this.detectAssetMode();
  }

  /**
   * Get singleton instance of AssetResolver
   */
  public static getInstance(): AssetResolver {
    if (!AssetResolver.instance) {
      AssetResolver.instance = new AssetResolver();
    }
    return AssetResolver.instance;
  }

  /**
   * Reset singleton instance (useful for testing)
   * @internal
   */
  public static resetInstance(): void {
    AssetResolver.instance = null;
  }

  /**
   * Detect current environment from NODE_ENV
   * @private
   */
  private detectEnvironment(): Environment {
    const nodeEnv = typeof process !== 'undefined'
      ? process.env.NODE_ENV?.toLowerCase()
      : undefined;

    if (nodeEnv === 'production') return 'production';
    if (nodeEnv === 'test') return 'test';
    return 'development';
  }

  /**
   * Detect asset mode from ASSET_MODE environment variable
   * @private
   */
  private detectAssetMode(): AssetMode {
    const assetMode = typeof process !== 'undefined'
      ? process.env.ASSET_MODE?.toLowerCase()
      : undefined;

    if (assetMode === 'local' || assetMode === 'cdn') {
      return assetMode as AssetMode;
    }
    return 'auto';
  }

  /**
   * Validate local path to prevent directory traversal
   * @private
   */
  private validateLocalPath(path: string): boolean {
    if (!path || typeof path !== 'string') return false;

    // Prevent directory traversal attacks
    if (path.includes('..')) return false;
    if (path.includes('\\')) return false;

    // Should start with / or ./
    if (!path.startsWith('/') && !path.startsWith('./')) return false;

    return true;
  }

  /**
   * Validate CDN URL format
   * @private
   */
  private validateCdnUrl(url: string): boolean {
    if (!url || typeof url !== 'string') return false;

    try {
      const parsed = new URL(url);

      // In production, require HTTPS
      if (this.environment === 'production' && parsed.protocol !== 'https:') {
        console.warn(`[AssetResolver] CDN URL should use HTTPS in production: ${url}`);
        return false;
      }

      return parsed.protocol === 'http:' || parsed.protocol === 'https:';
    } catch {
      return false;
    }
  }

  /**
   * Determine whether to use CDN based on environment and mode
   * @private
   */
  private shouldUseCdn(envMode: EnvMode = 'cdn-production-local-dev'): boolean {
    // Handle explicit asset mode override
    if (this.assetMode === 'cdn') return true;
    if (this.assetMode === 'local') return false;

    // Handle envMode strategies
    switch (envMode) {
      case 'cdn-always':
        return true;

      case 'local-always':
        return false;

      case 'cdn-production-local-dev':
      default:
        return this.environment === 'production';
    }
  }

  /**
   * Get environment-aware asset URL
   *
   * @param localPath - Local file path (e.g., '/media/logo.png')
   * @param cdnUrl - CDN URL (e.g., 'https://cdn.example.com/logos/logo.png')
   * @param envMode - Environment mode strategy (default: 'cdn-production-local-dev')
   * @returns Resolved asset URL based on environment
   *
   * @example
   * ```typescript
   * const resolver = AssetResolver.getInstance();
   *
   * // Default behavior: CDN in prod, local in dev
   * const url1 = resolver.getAssetUrl('/media/logo.png', 'https://cdn.example.com/logo.png');
   *
   * // Always use CDN
   * const url2 = resolver.getAssetUrl('/media/banner.jpg', 'https://cdn.example.com/banner.jpg', 'cdn-always');
   *
   * // Always use local
   * const url3 = resolver.getAssetUrl('/media/icon.svg', 'https://cdn.example.com/icon.svg', 'local-always');
   * ```
   */
  public getAssetUrl(
    localPath: string,
    cdnUrl: string,
    envMode: EnvMode = 'cdn-production-local-dev'
  ): string {
    // Validate inputs
    const isLocalValid = this.validateLocalPath(localPath);
    const isCdnValid = this.validateCdnUrl(cdnUrl);

    if (!isLocalValid && !isCdnValid) {
      throw new Error(`[AssetResolver] Invalid asset paths: local="${localPath}", cdn="${cdnUrl}"`);
    }

    // Determine which URL to use
    const useCdn = this.shouldUseCdn(envMode);

    if (useCdn && isCdnValid) {
      return cdnUrl;
    }

    if (!useCdn && isLocalValid) {
      return localPath;
    }

    // Fallback logic
    if (isLocalValid) {
      console.warn(`[AssetResolver] Falling back to local path: ${localPath}`);
      return localPath;
    }

    if (isCdnValid) {
      console.warn(`[AssetResolver] Falling back to CDN URL: ${cdnUrl}`);
      return cdnUrl;
    }

    throw new Error(`[AssetResolver] Cannot resolve asset URL`);
  }

  /**
   * Get current environment
   */
  public getEnvironment(): Environment {
    return this.environment;
  }

  /**
   * Get current asset mode
   */
  public getAssetMode(): AssetMode {
    return this.assetMode;
  }
}

/**
 * Convenience function for getting asset URL without creating resolver instance
 *
 * @param localPath - Local file path
 * @param cdnUrl - CDN URL
 * @param envMode - Environment mode strategy
 * @returns Resolved asset URL
 *
 * @example
 * ```typescript
 * import { getAssetUrl } from '@/lib/assets';
 *
 * const logoUrl = getAssetUrl('/media/logo.png', 'https://cdn.example.com/logo.png');
 * ```
 */
export function getAssetUrl(
  localPath: string,
  cdnUrl: string,
  envMode?: EnvMode
): string {
  const resolver = AssetResolver.getInstance();
  return resolver.getAssetUrl(localPath, cdnUrl, envMode);
}

/**
 * React hook for environment-aware asset URLs with memoization
 *
 * Automatically re-computes only when inputs change, preventing unnecessary re-renders.
 *
 * @param localPath - Local file path
 * @param cdnUrl - CDN URL
 * @param envMode - Environment mode strategy
 * @returns Resolved asset URL (memoized)
 *
 * @example
 * ```tsx
 * import { useAsset } from '@/lib/assets';
 *
 * export default function Logo() {
 *   const logoUrl = useAsset(
 *     '/media/logo.png',
 *     'https://cdn.example.com/logos/logo.png'
 *   );
 *
 *   return <img src={logoUrl} alt="Logo" />;
 * }
 * ```
 *
 * @example
 * ```tsx
 * // With custom env mode
 * function Banner() {
 *   const bannerUrl = useAsset(
 *     '/media/banner.jpg',
 *     'https://cdn.example.com/banners/hero.jpg',
 *     'cdn-always'
 *   );
 *
 *   return <div style={{ backgroundImage: `url(${bannerUrl})` }} />;
 * }
 * ```
 */
export function useAsset(
  localPath: string,
  cdnUrl: string,
  envMode?: EnvMode
): string {
  // Check if React hooks are available
  if (typeof require !== 'undefined') {
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const React = require('react');

      // eslint-disable-next-line react-hooks/rules-of-hooks
      return React.useMemo(
        () => getAssetUrl(localPath, cdnUrl, envMode),
        [localPath, cdnUrl, envMode]
      );
    } catch {
      // React not available, fall through to simple resolution
    }
  }

  // Fallback if React is not available
  return getAssetUrl(localPath, cdnUrl, envMode);
}

/**
 * Batch resolve multiple assets
 *
 * @param configs - Array of asset configurations
 * @returns Array of resolved URLs in same order
 *
 * @example
 * ```typescript
 * import { batchResolveAssets } from '@/lib/assets';
 *
 * const [logoUrl, bannerUrl, iconUrl] = batchResolveAssets([
 *   { localPath: '/media/logo.png', cdnUrl: 'https://cdn.example.com/logo.png' },
 *   { localPath: '/media/banner.jpg', cdnUrl: 'https://cdn.example.com/banner.jpg' },
 *   { localPath: '/media/icon.svg', cdnUrl: 'https://cdn.example.com/icon.svg', envMode: 'cdn-always' },
 * ]);
 * ```
 */
export function batchResolveAssets(configs: AssetConfig[]): string[] {
  const resolver = AssetResolver.getInstance();
  return configs.map(config =>
    resolver.getAssetUrl(config.localPath, config.cdnUrl, config.envMode)
  );
}

// Export types
export type { AssetConfig, Environment };

File: templates/project/lib/assets.py
"""
Environment-Aware Asset Helper for Python Projects

This utility provides automatic asset URL resolution based on environment:
- Development: Uses local file paths
- Production: Uses CDN URLs
- Test: Configurable behavior

Compatible with FastAPI, Flask, Django, and standalone Python applications.

Version: 1.0.0
License: MIT
"""

import os
import re
from enum import Enum
from functools import lru_cache
from typing import Literal, Optional, List, Dict
from urllib.parse import urlparse


class AssetMode(Enum):
    """
    Asset resolution mode

    - LOCAL: Always use local paths
    - CDN: Always use CDN URLs
    - AUTO: Automatic based on environment (default)
    """
    LOCAL = 'local'
    CDN = 'cdn'
    AUTO = 'auto'


# Type alias for environment-specific strategies
EnvMode = Literal['cdn-production-local-dev', 'cdn-always', 'local-always']


class AssetResolver:
    """
    AssetResolver - Singleton class for environment-aware asset URL resolution

    Provides cached environment detection and URL resolution logic.

    Examples:
        >>> resolver = AssetResolver.get_instance()
        >>> url = resolver.get_asset_url('/media/logo.png', 'https://cdn.example.com/logos/logo.png')
        >>> print(url)
        '/media/logo.png'  # in development
        'https://cdn.example.com/logos/logo.png'  # in production
    """

    _instance: Optional['AssetResolver'] = None

    def __init__(self):
        """Private constructor - use get_instance() instead"""
        if AssetResolver._instance is not None:
            raise RuntimeError("Use AssetResolver.get_instance() instead of constructor")

        self.environment = self._detect_environment()
        self.asset_mode = self._detect_asset_mode()

    @classmethod
    def get_instance(cls) -> 'AssetResolver':
        """Get singleton instance of AssetResolver"""
        if cls._instance is None:
            cls._instance = cls.__new__(cls)
            cls._instance.__init__()
        return cls._instance

    @classmethod
    def reset_instance(cls) -> None:
        """Reset singleton instance (useful for testing)"""
        cls._instance = None

    @lru_cache(maxsize=1)
    def _detect_environment(self) -> str:
        """
        Detect current environment from ENVIRONMENT or PYTHON_ENV variables

        Returns:
            'production', 'development', or 'test'
        """
        # Check ENVIRONMENT first (most common in Python)
        env = os.getenv('ENVIRONMENT', '').lower()
        if env in ('production', 'prod'):
            return 'production'
        if env in ('development', 'dev'):
            return 'development'
        if env == 'test':
            return 'test'

        # Fallback to PYTHON_ENV
        python_env = os.getenv('PYTHON_ENV', '').lower()
        if python_env in ('production', 'prod'):
            return 'production'
        if python_env in ('development', 'dev'):
            return 'development'
        if python_env == 'test':
            return 'test'

        # Default to development
        return 'development'

    @lru_cache(maxsize=1)
    def _detect_asset_mode(self) -> AssetMode:
        """
        Detect asset mode from ASSET_MODE environment variable

        Returns:
            AssetMode enum value
        """
        asset_mode = os.getenv('ASSET_MODE', '').lower()

        if asset_mode == 'local':
            return AssetMode.LOCAL
        if asset_mode == 'cdn':
            return AssetMode.CDN

        return AssetMode.AUTO

    def _validate_local_path(self, path: str) -> bool:
        """
        Validate local path to prevent directory traversal

        Args:
            path: Local file path

        Returns:
            True if valid, False otherwise
        """
        if not path or not isinstance(path, str):
            return False

        # Prevent directory traversal attacks
        if '..' in path:
            return False

        # Should start with / or ./
        if not path.startswith('/') and not path.startswith('./'):
            return False

        return True

    def _validate_cdn_url(self, url: str) -> bool:
        """
        Validate CDN URL format

        Args:
            url: CDN URL

        Returns:
            True if valid, False otherwise
        """
        if not url or not isinstance(url, str):
            return False

        try:
            parsed = urlparse(url)

            # In production, require HTTPS
            if self.environment == 'production' and parsed.scheme != 'https':
                print(f"[AssetResolver] Warning: CDN URL should use HTTPS in production: {url}")
                return False

            return parsed.scheme in ('http', 'https')
        except Exception:
            return False

    def _should_use_cdn(self, env_mode: EnvMode = 'cdn-production-local-dev') -> bool:
        """
        Determine whether to use CDN based on environment and mode

        Args:
            env_mode: Environment mode strategy

        Returns:
            True if CDN should be used, False otherwise
        """
        # Handle explicit asset mode override
        if self.asset_mode == AssetMode.CDN:
            return True
        if self.asset_mode == AssetMode.LOCAL:
            return False

        # Handle envMode strategies
        if env_mode == 'cdn-always':
            return True
        elif env_mode == 'local-always':
            return False
        else:  # cdn-production-local-dev (default)
            return self.environment == 'production'

    def get_asset_url(
        self,
        local_path: str,
        cdn_url: str,
        env_mode: EnvMode = 'cdn-production-local-dev'
    ) -> str:
        """
        Get environment-aware asset URL

        Args:
            local_path: Local file path (e.g., '/media/logo.png')
            cdn_url: CDN URL (e.g., 'https://cdn.example.com/logos/logo.png')
            env_mode: Environment mode strategy (default: 'cdn-production-local-dev')

        Returns:
            Resolved asset URL based on environment

        Raises:
            ValueError: If both paths are invalid

        Examples:
            >>> resolver = AssetResolver.get_instance()

            # Default behavior: CDN in prod, local in dev
            >>> url1 = resolver.get_asset_url('/media/logo.png', 'https://cdn.example.com/logo.png')

            # Always use CDN
            >>> url2 = resolver.get_asset_url(
            ...     '/media/banner.jpg',
            ...     'https://cdn.example.com/banner.jpg',
            ...     'cdn-always'
            ... )

            # Always use local
            >>> url3 = resolver.get_asset_url(
            ...     '/media/icon.svg',
            ...     'https://cdn.example.com/icon.svg',
            ...     'local-always'
            ... )
        """
        # Validate inputs
        is_local_valid = self._validate_local_path(local_path)
        is_cdn_valid = self._validate_cdn_url(cdn_url)

        if not is_local_valid and not is_cdn_valid:
            raise ValueError(
                f"[AssetResolver] Invalid asset paths: "
                f'local="{local_path}", cdn="{cdn_url}"'
            )

        # Determine which URL to use
        use_cdn = self._should_use_cdn(env_mode)

        if use_cdn and is_cdn_valid:
            return cdn_url

        if not use_cdn and is_local_valid:
            return local_path

        # Fallback logic
        if is_local_valid:
            print(f"[AssetResolver] Warning: Falling back to local path: {local_path}")
            return local_path

        if is_cdn_valid:
            print(f"[AssetResolver] Warning: Falling back to CDN URL: {cdn_url}")
            return cdn_url

        raise ValueError("[AssetResolver] Cannot resolve asset URL")

    def get_environment(self) -> str:
        """Get current environment"""
        return self.environment

    def get_asset_mode(self) -> AssetMode:
        """Get current asset mode"""
        return self.asset_mode


def get_asset_url(
    local_path: str,
    cdn_url: str,
    env_mode: EnvMode = 'cdn-production-local-dev'
) -> str:
    """
    Convenience function for getting asset URL without creating resolver instance

    Args:
        local_path: Local file path
        cdn_url: CDN URL
        env_mode: Environment mode strategy

    Returns:
        Resolved asset URL

    Examples:
        >>> from lib.assets import get_asset_url

        >>> logo_url = get_asset_url('/media/logo.png', 'https://cdn.example.com/logo.png')
        >>> print(logo_url)
        '/media/logo.png'  # in development

        >>> # FastAPI example
        >>> from fastapi import FastAPI
        >>> from lib.assets import get_asset_url

        >>> app = FastAPI()

        >>> @app.get("/")
        >>> def read_root():
        ...     return {
        ...         "logo": get_asset_url('/static/logo.png', 'https://cdn.example.com/logo.png')
        ...     }

        >>> # Flask example
        >>> from flask import Flask
        >>> from lib.assets import get_asset_url

        >>> app = Flask(__name__)

        >>> @app.route("/")
        >>> def index():
        ...     logo_url = get_asset_url('/static/logo.png', 'https://cdn.example.com/logo.png')
        ...     return f'<img src="{logo_url}" />'
    """
    resolver = AssetResolver.get_instance()
    return resolver.get_asset_url(local_path, cdn_url, env_mode)


def batch_resolve_assets(configs: List[Dict[str, str]]) -> List[str]:
    """
    Batch resolve multiple assets

    Args:
        configs: List of asset configurations with keys:
                - local_path: Local file path
                - cdn_url: CDN URL
                - env_mode: (optional) Environment mode strategy

    Returns:
        List of resolved URLs in same order

    Examples:
        >>> from lib.assets import batch_resolve_assets

        >>> urls = batch_resolve_assets([
        ...     {
        ...         'local_path': '/media/logo.png',
        ...         'cdn_url': 'https://cdn.example.com/logo.png'
        ...     },
        ...     {
        ...         'local_path': '/media/banner.jpg',
        ...         'cdn_url': 'https://cdn.example.com/banner.jpg'
        ...     },
        ...     {
        ...         'local_path': '/media/icon.svg',
        ...         'cdn_url': 'https://cdn.example.com/icon.svg',
        ...         'env_mode': 'cdn-always'
        ...     }
        ... ])
        >>> logo_url, banner_url, icon_url = urls
    """
    resolver = AssetResolver.get_instance()

    results = []
    for config in configs:
        local_path = config.get('local_path', '')
        cdn_url = config.get('cdn_url', '')
        env_mode = config.get('env_mode', 'cdn-production-local-dev')

        url = resolver.get_asset_url(local_path, cdn_url, env_mode)
        results.append(url)

    return results


# Convenience exports
__all__ = [
    'AssetMode',
    'EnvMode',
    'AssetResolver',
    'get_asset_url',
    'batch_resolve_assets',
]
